### PR TITLE
[db_migrator] Migrate DNS configuratuion

### PR DESCRIPTION
**Pull request by ganglyu**
#### What I did
https://github.com/sonic-net/sonic-buildimage/pull/14549
This PR has introduced DNS configuration to CONFIG_DB, we need to migrate original CONFIG_DB to support DNS configuration during upgrading SONiC.

#### How I did it
If there's no DNS_NAMESERVER table in CONFIG_DB, read DNS_NAMESERVER table from minigraph and update CONFIG_DB.

#### How to verify it
Run unit test for db migrator.

### CODE REVIEW

File: tests/db_migrator_test.py [hunk 1]
```diff
@@ -294,6 +294,29 @@ def test_lacp_key_migrator(self):
...
+
+    def test_dns_nameserver_migrator(self):
+        dbconnector.dedicated_dbs['CONFIG_DB'] = os.path.join(mock_db_path, 'config_db', 'dns-nameserver-input')
+        import db_migrator
+        dbmgtr = db_migrator.DBMigrator(None)
+        dbmgtr.migrate()
+        dbconnector.dedicated_dbs['CONFIG_DB'] = os.path.join(mock_db_path, 'config_db', 'dns-nameserver-expected')
+        expected_db = Db()
+        resulting_keys = dbmgtr.configDB.keys(dbmgtr.configDB.CONFIG_DB, 'DNS_NAMESERVER*')
+        expected_kesy = expected_db.cfgdb.keys(expected_db.cfgdb.CONFIG_DB, 'DNS_NAMESERVER*')
```
Comment (vaibhavhd) [inline tests/db_migrator_test.py:315]: Typo in variable name `expected_kesy`

File: tests/db_migrator_input/config_db/dns-nameserver-expected.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+    "DNS_NAMESERVER|1.1.1.1": {
+        "state": "enabled"
+    },
+    "DNS_NAMESERVER|8.8.8.8": {
```
Comment (vaibhavhd) [inline tests/db_migrator_input/config_db/dns-nameserver-expected.json:5]: It might be a good idea to add a test entry for IPv6 nameserver.

**Review by vaibhavhd — APPROVED** [review]
Non blocking comments added.

**Comment by qiluo-msft** [comment]
Did you manually tested the workflow of https://github.com/sonic-net/SONiC/wiki/L2-Switch-mode?

File: tests/db_migrator_input/minigraph.xml [hunk 3]
```diff
@@ -19,9 +19,17 @@
...
-    <Devices xmlns:a="http://schemas">
+    <Devices xmlns:a="http://schemas.datacontract.org/2004/07/Microsoft.Search.Autopilot.Evolution">
       <a:DeviceMetadata>
-        <a:Properties/>
+        <a:Name>SONiC-Dummy</a:Name>
+        <a:Properties>
+          <a:DeviceProperty>
+            <a:Name>DnsNameserverResources</a:Name>
+            <a:Value>
+             1.1.1.1;8.8.8.8
```
Comment (qiluo-msft) [inline tests/db_migrator_input/minigraph.xml:29]: >1.1.1.1;8.8.8.8 [](http://example.com/codeflow?start=13&length=15)

Did you test empty Value?

File: tests/db_migrator_input/minigraph.xml [hunk 4]
```diff
@@ -19,9 +19,17 @@
...
    </PngDec>
   <MetadataDeclaration>
-    <Devices xmlns:a="http://schemas">
+    <Devices xmlns:a="http://schemas.datacontract.org/2004/07/Microsoft.Search.Autopilot.Evolution">
       <a:DeviceMetadata>
-        <a:Properties/>
+        <a:Name>SONiC-Dummy</a:Name>
+        <a:Properties>
+          <a:DeviceProperty>
+            <a:Name>DnsNameserverResources</a:Name>
```
Comment (qiluo-msft) [inline tests/db_migrator_input/minigraph.xml:27]: >DnsNameserverResources [](http://example.com/codeflow?start=20&length=22)

Block this PR until minigraph schema is accepted by our internal team.

File: scripts/db_migrator.py [hunk 5]
```diff
@@ -636,6 +636,18 @@ def migrate_route_table(self):
             if 'protocol' not in route_attr:
                 self.appDB.set(self.appDB.APPL_DB, route_key, 'protocol', '')
 
+    def migrate_dns_nameserver(self):
```
Comment (qiluo-msft) [inline scripts/db_migrator.py:639]: Will you create a new DBMigrator version? If no, I guess there will be overhead to warm reboot.

File: tests/db_migrator_input/minigraph.xml (hunk 4 above)
Comment (ganglyu) [reply tests/db_migrator_input/minigraph.xml:27]: I think minigraph schema has been accepted.

File: tests/db_migrator_input/minigraph.xml (hunk 3 above)
Comment (qiluo-msft) [reply tests/db_migrator_input/minigraph.xml:29]: Actually, we should revert minigraph schema code change and test cases.

File: tests/db_migrator_input/minigraph.xml (hunk 4 above)
Comment (qiluo-msft) [reply tests/db_migrator_input/minigraph.xml:27]: Actually, we should revert minigraph schema code change and test cases.

File: scripts/db_migrator.py (hunk 5 above)
Comment (ganglyu) [reply scripts/db_migrator.py:639]: Now I use the latest version: version_4_0_3

**Review by qiluo-msft — DISMISSED** [review]
LGTM. Please check with other reviewers.

File: scripts/db_migrator.py [hunk 6]
```diff
@@ -1007,6 +1019,8 @@ def version_4_0_3(self):
         This is the latest version for master branch
         """
         log.log_info('Handling version_4_0_3')
+        # Updating DNS nameserver
+        self.migrate_dns_nameserver()
```
Comment (vaibhavhd) [inline scripts/db_migrator.py:1023]: You updated the existing version, but did not create a new version.

Please add new version handler here as a placeholder. 

Also, when you do that please also update the self.CURRENT_VERSION to the new version.

File: tests/db_migrator_test.py [hunk 7]
```diff
@@ -294,6 +294,30 @@ def test_lacp_key_migrator(self):
...
+        os.environ['UTILITIES_UNIT_TESTING'] = "0"
+        dbconnector.dedicated_dbs['CONFIG_DB'] = None
+
+    def test_dns_nameserver_migrator(self):
+        dbconnector.dedicated_dbs['CONFIG_DB'] = os.path.join(mock_db_path, 'config_db', 'dns-nameserver-input')
+        import db_migrator
+        dbmgtr = db_migrator.DBMigrator(None)
+        dbmgtr.migrate()
+        dbconnector.dedicated_dbs['CONFIG_DB'] = os.path.join(mock_db_path, 'config_db', 'dns-nameserver-expected')
+        expected_db = Db()
```
Comment (vaibhavhd) [inline tests/db_migrator_test.py:313]: You have removed the case where config migration was tested w/ entries from minigraph.xml . 

I think we should still have that case.

File: tests/db_migrator_test.py [hunk 8]
```diff
@@ -294,6 +294,30 @@ def test_lacp_key_migrator(self):
...
+        advance_version_for_expected_database(dbmgtr.configDB, expected_db.cfgdb, 'version_4_0_3')
```
Comment (qiluo-msft) [inline tests/db_migrator_test.py:314]: >expected_db [](http://example.com/codeflow?start=63&length=11)

The expected_db should not be empty.

**Comment by vaibhavhd** [comment]
Please do not merge this change until https://github.com/sonic-net/sonic-utilities/pull/2895 is merged.

2895 fixes an issue where CURRENT_VERSION is not updated.

File: tests/db_migrator_test.py (hunk 8 above)
Comment (ganglyu) [reply tests/db_migrator_test.py:314]: We have a PR to revert minigraph schema code change, and minigraph.py for public will not generate any DNS configuration, that's why the expected_db is empty.

Comment (qiluo-msft) [reply tests/db_migrator_test.py:314]: Then this PR is blocked on that PR

File: tests/db_migrator_test.py (hunk 1 above)
Comment (ganglyu) [reply tests/db_migrator_test.py]: Fixed

File: tests/db_migrator_input/minigraph.xml (hunk 3 above)
Comment (ganglyu) [reply tests/db_migrator_input/minigraph.xml]: https://github.com/sonic-net/sonic-buildimage/pull/15727

File: scripts/db_migrator.py (hunk 5 above)
Comment (ganglyu) [reply scripts/db_migrator.py:639]: Use version_4_0_4

File: scripts/db_migrator.py (hunk 6 above)
Comment (ganglyu) [reply scripts/db_migrator.py:1031]: Updated

File: tests/db_migrator_test.py (hunk 8 above)
Comment (ganglyu) [reply tests/db_migrator_test.py]: Use mocked minigraph data

File: tests/db_migrator_test.py (hunk 7 above)
Comment (ganglyu) [reply tests/db_migrator_test.py:320]: Fixed

File: tests/db_migrator_input/config_db/dns-nameserver-expected.json (hunk 2 above)
Comment (ganglyu) [reply tests/db_migrator_input/config_db/dns-nameserver-expected.json]: Updated.

**Comment by qiluo-msft** [comment]
Please resolve the conflict.

**Review by qiluo-msft — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Comment by StormLiangMS** [comment]
@ganglyu do we have ADO for this?

**Comment by StormLiangMS** [comment]
@ganglyu is this an enhancement? If yes, no for backport to 202305

**Comment by dgsudharsan** [comment]
Removing the cherry-pick tag since the feature of DNS is not part of 202305

**Pull request merged**